### PR TITLE
Grouped conformal predictors

### PR DESCRIPTION
**Pull request by paxcema**
## Why
Time series tasks with grouped-by columns had poor confidence estimation due to a single Inductive Conformal Predictor handling all data at once. For informative prediction bounds, we need to instance one ICP per each group combination.

Related issue: [M#1116](https://github.com/mindsdb/mindsdb/issues/1116)

## How
To avoid unnecessary reshaping, this PR generates and caches prediction results inside `LightwoodBackend.Predict()`

Regarding confidence estimation, helpers have been divided into conformal and generic confidence helpers. The procedures to 1) calibrate ICPs in Model Analysis and 2) to estimate query confidence in PredictTransaction, have been restructured to handle the case of grouped time series.

### Notes
~~This is staying as a draft while benchmarks are finished and to clean up the code~~.

### CODE REVIEW

File: mindsdb_native/libs/phases/model_interface/lightwood_backend.py [hunk 1]
```diff
@@ -536,6 +536,16 @@ def predict(self, mode='predict', ignore_columns=None, all_mixers=False):
 
             predictions = self.predictor.predict(when_data=run_df)
 
+            # cache run_df to avoid duplicate reshaping in analysis and predict transactions
+            if mode == 'predict_on_train_data':
+                self.transaction.input_data.cached_train_df = run_df
```
Comment (George3d6) [inline mindsdb_native/libs/phases/model_interface/lightwood_backend.py]: cache that during the training phase.

File: mindsdb_native/libs/phases/model_interface/lightwood_backend.py [hunk 2]
```diff
@@ -536,6 +536,16 @@ def predict(self, mode='predict', ignore_columns=None, all_mixers=False):
...
+            elif mode == 'validate':
+                self.transaction.input_data.cached_val_df = run_df
+            elif mode == 'test':
+                self.transaction.input_data.cached_test_df = run_df
```
Comment (George3d6) [inline mindsdb_native/libs/phases/model_interface/lightwood_backend.py]: cache that during the training phase.

File: mindsdb_native/libs/phases/model_analyzer/model_analyzer.py [hunk 3]
```diff
@@ -92,55 +100,108 @@ def run(self):
...
+
+                    icps = self.transaction.hmd['icp'][target]
+                    group_keys = icps['__group_keys']
+                    result_df = pd.DataFrame(index=self.transaction.input_data.cached_val_df.index,
+                                             columns=['lower', 'upper'])
+
+                    for group in icps['__groups']:
+                        icp_df = deepcopy(self.transaction.input_data.cached_val_df)
+
+                        # add all predictions to the DF
```
Comment (paxcema) [inline mindsdb_native/libs/phases/model_analyzer/model_analyzer.py]: move this up

File: mindsdb_native/libs/phases/model_analyzer/model_analyzer.py [hunk 4]
```diff
@@ -92,55 +100,108 @@ def run(self):
...
+                        if is_multi_ts:
+                            icp_df[f'__predicted_{target}'] = [p[0] for p in normal_predictions[target]]
+                        else:
+                            icp_df[f'__predicted_{target}'] = normal_predictions[target]
```
Comment (George3d6) [inline mindsdb_native/libs/phases/model_analyzer/model_analyzer.py]: move before the for.

File: mindsdb_native/libs/phases/model_analyzer/model_analyzer.py [hunk 5]
```diff
@@ -92,55 +100,108 @@ def run(self):
...
+
+                        # filter irrelevant rows for each group combination
+                        for key, val in zip(group_keys, group):
```
Comment (paxcema) [inline mindsdb_native/libs/phases/model_analyzer/model_analyzer.py:172]: check if mutable, might not be necessary

File: mindsdb_native/libs/phases/model_analyzer/model_analyzer.py [hunk 6]
```diff
@@ -92,55 +100,108 @@ def run(self):
...
+                            self.transaction.lmd['stats_v2'][target]['train_std_dev'][frozenset(group)] = y_train.std()
+
+                        # get bounds for relevant rows in validation dataset
+                        _, group_ranges = set_conf_range(icp_df, icps[frozenset(group)], target, typing_info,
+                                                                  self.transaction.lmd, group=group)
+                        result_df['lower'][icp_df.index] = group_ranges[:, 0]
+                        result_df['upper'][icp_df.index] = group_ranges[:, 1]
+
+                    # consolidate all groups here
+                    ranges = result_df.values
```
Comment (George3d6) [inline mindsdb_native/libs/phases/model_analyzer/model_analyzer.py]: Throw out these and also the code that prepares sample ts data for scout.

**Review by George3d6 — CHANGES_REQUESTED** [review]

**Comment by paxcema** [comment]
- [x] Handle non-grouped case with a `__default` key to streamline the code in `ModelAnalysis` and `PredictTransaction`

**Pull request merged**